File: packages/browser/src/site-runtime.ts
import { validateSiteResources } from "./site-resource";
import {
  canonical,
  exactShape,
  type Bundle,
  type PublicIdentity,
} from "../../core/src/protocol";
import { validateContentShape } from "../../content/src/validation";
import type { Content, DisplayObject } from "../../content/src/types";
import { parseSiteAddress, siteAddress } from "../../sites/src/protocol";
import { BrowserProfile } from "./profile";
import { BrowserSiteCatalog, type SiteOperationHandle } from "./site-catalog";
import { verifySiteContent } from "./site-content";
type Operation = NonNullable<
  Awaited<ReturnType<BrowserSiteCatalog["operation"]>>
>;
interface Context {
  profile: BrowserProfile;
  blocked(): Promise<readonly string[]>;
  knownWithdrawals?(
    readValue: (key: string) => Promise<unknown>,
  ): Promise<Record<string, string>>;
  readers(ids: unknown): Promise<PublicIdentity[] | "public">;
  publish(bundle: Bundle): void;
  request(ids: string[]): Promise<void>;
}
export class BrowserSiteRuntime {
  private readonly catalog: BrowserSiteCatalog;
  private readonly owner: PublicIdentity;
  private readonly generation: number;
  private initialized = false;
  private closed = false;
  private tail = Promise.resolve();
  private pending = new Map<string, { name: string; operation: Operation }>();
  private nextRetry = 0;
  private ticking = false;
  lastError = "";
  constructor(private readonly context: Context) {
    const owner = context.profile.identity;
    if (!owner) throw new Error("Desbloqueia a identidade");
    this.owner = owner;
    this.generation = context.profile.sessionGeneration;
    this.catalog = new BrowserSiteCatalog(context.profile);
  }
  private ensure() {
    if (
      this.closed ||
      this.context.profile.sessionGeneration !== this.generation
    )
      throw new Error("Sessão de site bloqueada");
  }
  close() {
    this.closed = true;
    this.pending.clear();
  }
  private serial<T>(run: () => Promise<T>): Promise<T> {
    const work = this.tail.then(async () => {
      this.ensure();
      const result = await run();
      this.ensure();
      return result;
    });
    this.tail = work.then(
      () => {},
      () => {},
    );
    return work;
  }
  private slot(name: string, handle: SiteOperationHandle) {
    return name + ":" + handle.sequence;
  }
  private async inspect(bundle: Bundle) {
    if (bundle.manifest.kind !== "site") return null;
    if (
      bundle.manifest.publicKey === null &&
      !bundle.manifest.keys.some((k) => k.reader === this.owner.id)
    )
      return null;
    const content = (await this.context.profile.decrypt(bundle)) as Content;
    validateContentShape(content);
    if (content.type !== "site") throw new Error("Conteúdo de site inválido");
    return verifySiteContent(content, bundle.manifest.author);
  }
  private async initialize() {
    if (this.initialized) return;
    const blocked = await this.context.blocked();
    for (const id of await this.context.profile.ids()) {
      let bundle: Bundle, parsed: ReturnType<typeof verifySiteContent>;
      try {
        bundle = await this.context.profile.getBundle(id);
        if (
          bundle.manifest.kind !== "site" ||
          blocked.includes(bundle.manifest.author.id)
        )
          continue;
        parsed = await this.inspect(bundle);
      } catch {
        continue;
      }
      if (parsed) await this.catalog.observe(bundle, parsed.revision.body.name);
    }
    this.pending.clear();
    for (const item of await this.catalog.pending())
      this.pending.set(this.slot(item.name, item), {
        name: item.name,
        operation: item,
      });
    this.initialized = true;
  }
  receive(bundle: Bundle) {
    return this.serial(async () => {
      const parsed = await this.inspect(bundle);
      if (parsed) await this.catalog.observe(bundle, parsed.revision.body.name);
    });
  }
  private async current(name: string, h: SiteOperationHandle) {
    const op = await this.catalog.operation(name, h.sequence, h.operationId);
    if (!op || op.fingerprint !== h.fingerprint)
      throw new Error("Operação de site desconhecida");
    return op;
  }
  private async finish(name: string, h: SiteOperationHandle) {
    let operation = await this.current(name, h);
    const access = await this.catalog.preparationAccess(name, operation),
      blocked = await this.context.blocked();
    if (
      access?.readers !== "public" &&
      access?.readers.some((id) => blocked.includes(id))
    ) {
      this.lastError =
        "A publicação está em pausa porque um leitor foi bloqueado.";
      return { operation, error: this.lastError };
    }
    if (operation.phase === "prepared")
      operation = await this.catalog.commit(name, operation);
    if (operation.phase === "committed") {
      const bundle = await this.catalog.authorizedBundle(name, operation);
      this.ensure();
      await this.context.profile.putBundle(bundle, true);
      const stored = await this.context.profile.getBundle(bundle.manifest.id);
      if (canonical(stored) !== canonical(bundle))
        throw new Error("A cópia do site não corresponde à publicação");
      operation = await this.catalog.markReady(name, operation);
      this.ensure();
      this.context.publish(stored);
    }
    if (!["prepared", "committed"].includes(operation.phase))
      this.pending.delete(this.slot(name, operation));
    this.lastError = "";
    return { operation };
  }
  status() {
    return { pending: this.pending.size, error: this.lastError };
  }
  async tick() {
    if (this.closed || this.ticking || Date.now() < this.nextRetry) return;
    this.nextRetry = Date.now() + 5000;
    this.ticking = true;
    try {
      await this.serial(async () => {
        await this.initialize();
        const next = [...this.pending.values()].find(
          (p) => p.operation.phase === "committed" || p.operation.requested,
        );
        if (next) {
          const key = this.slot(next.name, next.operation);
          this.pending.delete(key);
          this.pending.set(key, next);
          const result = await this.finish(next.name, next.operation);
          this.lastError = result.error ?? "";
        }
      });
    } catch (error) {
      this.lastError = (error as Error).message;
    } finally {
      this.ticking = false;
    }
  }
  private async address(value: unknown) {
    const parsed = parseSiteAddress(value);
    if ((await this.context.blocked()).includes(parsed.ownerId))
      throw new Error("Contacto bloqueado");
    return parsed;
  }
  private async resolve(address: string, revisionId?: string) {
    const { ownerId, name } = await this.address(address),
      state = await this.catalog.state(ownerId, name);
    if (revisionId !== undefined && !/^[a-f0-9]{64}$/.test(revisionId))
      throw new Error("Revisão inválida");
    if (!revisionId && state.status === "conflict")
      return { status: "conflict", address, state };
    const chosen = revisionId ?? state.heads[0]?.id;
    if (!chosen) return { status: "pending", address, state };
    const header = (await this.catalog.history(ownerId, name)).find(
      (h) => h.revision.id === chosen,
    );
    if (!header)
      return { status: "unknown-revision", address, revisionId: chosen, state };
    const ids = [
      ...new Set([...header.bundles, ...(await this.context.profile.ids())]),
    ];
    for (const id of ids) {
      try {
        const bundle = await this.context.profile.getBundle(id);
        if (
          bundle.manifest.kind !== "site" ||
          bundle.manifest.author.id !== ownerId
        )
          continue;
        const parsed = await this.inspect(bundle);
        if (
          !parsed ||
          parsed.revision.id !== chosen ||
          parsed.revision.body.name !== name
        )
          continue;
        const records = await this.context.profile.records();
        const object: DisplayObject = {
          id,
          author: bundle.manifest.author,
          kind: "site",
          created: bundle.manifest.created,
          expires: bundle.manifest.expires,
          content: parsed.content,
          pinned: records[id]?.pinned === true,
          readers: bundle.manifest.keys.map((k) => k.reader),
          public: bundle.manifest.publicKey !== null,
        };
        return {
          status: "available",
          address,
          revisionId: chosen,
          object,
          state,
        };
      } catch {
        /* A remembered header does not prove current payload availability. */
      }
    }
    await this.context.request(header.bundles);
    return { status: "pending", address, revisionId: chosen, state };
  }
  command(value: any): Promise<any> {
    return this.serial(() => this.execute(value));
  }
  private async execute(value: any): Promise<any> {
    if (!value || typeof value !== "object" || Array.isArray(value))
      throw new Error("Comando de site inválido");
    if (value.action === "operation" || value.action === "cancel") {
      if (!exactShape(value, ["action", "name", "sequence", "operationId"]))
        throw new Error("Comando de site inválido");
      const operation = await this.catalog.operation(
        value.name,
        value.sequence,
        value.operationId,
      );
      if (value.action === "operation" || !operation) return { operation };
      const cancelled = await this.catalog.cancel(value.name, operation);
      if (cancelled.phase === "cancelled")
        this.pending.delete(this.slot(value.name, cancelled));
      return { operation: cancelled };
    }
    await this.initialize();
    if (value.action === "state" || value.action === "history") {
      if (!exactShape(value, ["action", "address"]))
        throw new Error("Comando de site inválido");
      const { ownerId, name } = await this.address(value.address);
      return value.action === "state"
        ? this.catalog.state(ownerId, name)
        : {
            address: value.address,
            revisions: await this.catalog.history(ownerId, name),
          };
    }
    if (value.action === "resolve") {
      if (
        !exactShape(value, ["action", "address"]) &&
        !exactShape(value, ["action", "address", "revisionId"])
      )
        throw new Error("Comando de site inválido");
      return this.resolve(value.address, value.revisionId);
    }
    if (value.action === "resume") {
      if (!exactShape(value, ["action", "name", "sequence", "operationId"]))
        throw new Error("Comando de site inválido");
      const operation = await this.catalog.operation(
        value.name,
        value.sequence,
        value.operationId,
      );
      if (!operation) throw new Error("Operação de site desconhecida");
      return this.finish(value.name, operation);
    }
    const fields = [
      "action",
      "name",
      "sequence",
      "operationId",
      "expectedBase",
      "payload",
      "recipients",
      "ttlMs",
    ];
    if (
      value.action !== "publish" ||
      (!exactShape(value, fields) &&
        !exactShape(value, [...fields, "confirmedHeads"]))
    )
      throw new Error("Comando de site inválido");
    const operation = await this.catalog.createPublication(
      value.name,
      {
        sequence: value.sequence,
        operationId: value.operationId,
        expectedBase: value.expectedBase,
        payload: value.payload,
        readers: await this.context.readers(value.recipients),
        ttlMs: value.ttlMs,
        ...(value.confirmedHeads !== undefined
          ? { confirmedHeads: value.confirmedHeads }
          : {}),
      },
      async (request, readValue) => {
        const settings = (await readValue("mesh-settings")) as any;
        const withdrawn =
          (await this.context.knownWithdrawals?.(readValue)) ?? {};
        return validateSiteResources(
          request,
          this.context.profile,
          settings?.blocked ?? [],
          withdrawn,
        );
      },
    );
    const address = siteAddress(this.owner.id, value.name);
    if (["prepared", "committed"].includes(operation.phase))
      this.pending.set(this.slot(value.name, operation), {
        name: value.name,
        operation,
      });
    try {
      return { address, ...(await this.finish(value.name, operation)) };
    } catch (error) {
      this.lastError = (error as Error).message;
      try {
        return {
          address,
          operation: await this.current(value.name, operation),
          error: this.lastError,
        };
      } catch {
        throw error;
      }
    }
  }
}

File: native/app/site_runtime.go
package app

import (
	"errors"
	"github.com/JohnnyPBelo/relayloom/native/core"
	"github.com/JohnnyPBelo/relayloom/native/groupstore"
	"github.com/JohnnyPBelo/relayloom/native/profilestate"
	"github.com/JohnnyPBelo/relayloom/native/sites"
	"github.com/JohnnyPBelo/relayloom/native/transport"
	"time"
)

func inspectSiteBundle(bundle core.Bundle, identity *core.Identity) (*sites.VerifiedContent, error) {
	if bundle.Manifest.Kind != "site" {
		return nil, nil
	}
	readable := bundle.Manifest.PublicKey != nil
	for _, key := range bundle.Manifest.Keys {
		if identity != nil && key.Reader == identity.Public.ID {
			readable = true
		}
	}
	if !readable {
		return nil, nil
	}
	value, err := core.DecryptBundle(bundle, identity)
	if err != nil {
		return nil, err
	}
	m, err := object(value)
	if err != nil {
		return nil, err
	}
	if text(m["type"]) != "site" {
		return nil, errors.New("conteúdo de site inválido")
	}
	if _, versioned := m["siteRevision"]; !versioned {
		doc, _ := m["site"].(map[string]any)
		version, _ := number(doc["version"])
		if version == 3 {
			return nil, errors.New("sites versão 3 exigem um snapshot assinado")
		}
		return nil, validateContent(Content(m))
	}
	verified, err := sites.VerifyContent(m, bundle.Manifest.Author, "")
	if err != nil {
		return nil, err
	}
	return &verified, nil
}

type siteDatabase func(func(*groupstore.Tx) error) error

func (d siteDatabase) Update(fn func(*groupstore.Tx) error) error { return d(fn) }

type pendingSite struct {
	name      string
	operation sites.OperationSummary
}
type siteRuntime struct {
	node        *Node
	owner       *core.Identity
	catalog     *sites.Catalog
	initialized bool
	pending     []pendingSite
	nextRetry   int64
	lastError   string
}

func (n *Node) siteDatabaseLocked(owner *core.Identity) siteDatabase {
	return siteDatabase(func(fn func(*groupstore.Tx) error) error {
		if n.identity != owner || n.privateDatabase == nil {
			return errors.New("sessão de site bloqueada")
		}
		err := n.privateDatabase.Update(func(tx *groupstore.Tx) error {
			state, e := profilestate.Read(tx)
			if e != nil {
				return e
			}
			if state == nil || state.Digest != n.privateDigest {
				return errors.New("estado privado desactualizado")
			}
			return fn(tx)
		})
		if errors.Is(err, groupstore.ErrIntegrity) || errors.Is(err, groupstore.ErrUnavailable) {
			_ = n.lockPrivateLocked()
		}
		return err
	})
}

func (n *Node) sitesLocked() (*siteRuntime, error) {
	if n.identity == nil || n.privateDatabase == nil {
		return nil, errors.New("desbloqueie a identidade")
	}
	if n.siteRuntime != nil && n.siteRuntime.owner == n.identity {
		return n.siteRuntime, nil
	}
	owner := n.identity
	database := n.siteDatabaseLocked(owner)
	n.siteRuntime = &siteRuntime{node: n, owner: owner, catalog: sites.NewCatalog(database, *owner), pending: []pendingSite{}}
	return n.siteRuntime, nil
}
func (n *Node) siteStatusLocked() any {
	if n.identity == nil {
		return nil
	}
	if n.siteRuntime == nil {
		return map[string]any{"pending": 0, "error": ""}
	}
	return map[string]any{"pending": len(n.siteRuntime.pending), "error": n.siteRuntime.lastError}
}
func (r *siteRuntime) track(name string, op sites.OperationSummary) {
	for i, p := range r.pending {
		if p.name == name && p.operation.Sequence == op.Sequence {
			r.pending[i] = pendingSite{name, op}
			return
		}
	}
	r.pending = append(r.pending, pendingSite{name, op})
}
func (r *siteRuntime) remove(name string, seq int64) {
	next := r.pending[:0]
	for _, p := range r.pending {
		if p.name != name || p.operation.Sequence != seq {
			next = append(next, p)
		}
	}
	r.pending = next
}
func (r *siteRuntime) initialize() error {
	if r.initialized {
		return nil
	}
	for _, m := range r.node.Store.List() {
		if m.Kind != "site" || contains(r.node.config.Blocked, m.Author.ID) {
			continue
		}
		bundle, err := r.node.Store.GetWithTouch(m.ID, false)
		if err != nil {
			continue
		}
		parsed, err := inspectSiteBundle(bundle, r.owner)
		if err != nil {
			continue
		}
		if parsed != nil {
			if _, err = r.catalog.Observe(bundle, parsed.Revision.Body.Name); err != nil {
				return err
			}
		}
	}
	pending, err := r.catalog.Pending()
	if err != nil {
		return err
	}
	r.pending = []pendingSite{}
	for _, p := range pending {
		r.track(p.Name, p.OperationSummary)
	}
	r.initialized = true
	return nil
}
func (r *siteRuntime) receive(bundle core.Bundle) error {
	parsed, err := inspectSiteBundle(bundle, r.owner)
	if err != nil {
		return err
	}
	if parsed != nil {
		_, err = r.catalog.Observe(bundle, parsed.Revision.Body.Name)
	}
	return err
}
func (r *siteRuntime) current(name string, h sites.OperationHandle) (sites.OperationSummary, error) {
	op, err := r.catalog.Operation(name, h.Sequence, h.OperationID)
	if err != nil {
		return sites.OperationSummary{}, err
	}
	if op == nil || op.Fingerprint != h.Fingerprint {
		return sites.OperationSummary{}, errors.New("operação de site desconhecida")
	}
	return *op, nil
}
func (r *siteRuntime) finish(name string, h sites.OperationHandle) (map[string]any, error) {
	op, err := r.current(name, h)
	if err != nil {
		return nil, err
	}
	access, err := r.catalog.PreparationAccess(name, op.OperationHandle)
	if err != nil {
		return nil, err
	}
	if access != nil {
		if readers, ok := access.Readers.([]string); ok {
			for _, id := range readers {
				if contains(r.node.config.Blocked, id) {
					r.lastError = "A publicação está em pausa porque um leitor foi bloqueado."
					return map[string]any{"operation": op, "error": r.lastError}, nil
				}
			}
		}
	}
	if op.Phase == "prepared" {
		op, err = r.catalog.Commit(name, op.OperationHandle)
		if err != nil {
			return nil, err
		}
	}
	if op.Phase == "committed" {
		bundle, e := r.catalog.AuthorizedBundle(name, op.OperationHandle)
		if e != nil {
			return nil, e
		}
		if _, e = r.node.Store.Put(bundle, true); e != nil {
			return nil, e
		}
		stored, e := r.node.Store.GetWithTouch(bundle.Manifest.ID, false)
		if e != nil {
			return nil, e
		}
		a, e := core.Canonical(stored)
		if e != nil {
			return nil, e
		}
		b, _ := core.Canonical(bundle)
		if string(a) != string(b) {
			return nil, errors.New("cópia do site não corresponde à publicação")
		}
		op, e = r.catalog.MarkReady(name, op.OperationHandle)
		if e != nil {
			return nil, e
		}
		allowed, e := r.node.maySeedLocked(stored.Manifest)
		if e != nil {
			return nil, e
		}
		if !allowed {
			return nil, errors.New("partilha de site interrompida")
		}
		if _, e = r.node.Router.Broadcast(map[string]any{"type": "bundle", "bundle": stored}, transport.Bulk, 2*time.Minute, false); e != nil {
			return nil, e
		}
	}
	if op.Phase != "prepared" && op.Phase != "committed" {
		r.remove(name, op.Sequence)
	}
	r.lastError = ""
	return map[string]any{"operation": op}, nil
}
func (r *siteRuntime) tick() {
	now := time.Now().UnixMilli()
	if now < r.nextRetry {
		return
	}
	r.nextRetry = now + 5000
	if err := r.initialize(); err != nil {
		r.lastError = err.Error()
		return
	}
	for i, p := range r.pending {
		if p.operation.Phase == "committed" || p.operation.Requested {
			r.pending = append(append(r.pending[:i:i], r.pending[i+1:]...), p)
			result, err := r.finish(p.name, p.operation.OperationHandle)
			if err != nil {
				r.lastError = err.Error()
			} else {
				r.lastError = text(result["error"])
			}
			return
		}
	}
}
func (r *siteRuntime) address(value any) (string, string, error) {
	owner, name, err := sites.ParseAddress(text(value))
	if err != nil {
		return "", "", err
	}
	if contains(r.node.config.Blocked, owner) {
		return "", "", errors.New("contacto bloqueado")
	}
	return owner, name, nil
}
func (r *siteRuntime) resolve(address, revisionID string, hasRevision bool) (map[string]any, error) {
	owner, name, err := r.address(address)
	if err != nil {
		return nil, err
	}
	state, err := r.catalog.State(owner, name)
	if err != nil {
		return nil, err
	}
	if hasRevision && !core.ValidAddress(revisionID) {
		return nil, errors.New("revisão inválida")
	}
	if !hasRevision && state.Status == "conflict" {
		return map[string]any{"status": "conflict", "address": address, "state": state}, nil
	}
	if !hasRevision && len(state.Heads) > 0 {
		revisionID = state.Heads[0].ID
	}
	if revisionID == "" {
		return map[string]any{"status": "pending", "address": address, "state": state}, nil
	}
	history, err := r.catalog.History(owner, name)
	if err != nil {
		return nil, err
	}
	var header *sites.StoredRevision
	for _, h := range history {
		if h.Revision.ID == revisionID {
			copy := h
			header = &copy
			break
		}
	}
	if header == nil {
		return map[string]any{"status": "unknown-revision", "address": address, "revisionId": revisionID, "state": state}, nil
	}
	candidates := append([]string{}, header.Bundles...)
	for _, m := range r.node.Store.List() {
		if m.Kind == "site" && m.Author.ID == owner && !contains(candidates, m.ID) {
			candidates = append(candidates, m.ID)
		}
	}
	for _, id := range candidates {
		bundle, e := r.node.Store.Get(id)
		if e != nil {
			continue
		}
		parsed, e := inspectSiteBundle(bundle, r.owner)
		if e != nil || parsed == nil || parsed.Revision.ID != revisionID || parsed.Revision.Body.Name != name || bundle.Manifest.Author.ID != owner {
			continue
		}
		object, e := r.node.displayLocked(bundle)
		if e != nil {
			continue
		}
		return map[string]any{"status": "available", "address": address, "revisionId": revisionID, "object": object, "state": state}, nil
	}
	now := time.Now().UnixMilli()
	wanted := []string{}
	for _, id := range header.Bundles {
		if now-r.node.requests[id] > 5000 {
			wanted = append(wanted, id)
			r.node.rememberRequestLocked(id, now)
		}
	}
	if len(wanted) > 0 {
		_, _ = r.node.Router.Broadcast(map[string]any{"type": "request", "ids": wanted}, transport.Normal, 2*time.Minute, false)
	}
	return map[string]any{"status": "pending", "address": address, "revisionId": revisionID, "state": state}, nil
}
func commandShape(body map[string]any, fields []string, optional string) bool {
	if len(body) != len(fields) && (optional == "" || len(body) != len(fields)+1) {
		return false
	}
	for _, key := range fields {
		if _, ok := body[key]; !ok {
			return false
		}
	}
	for key := range body {
		if key != optional && !contains(fields, key) {
			return false
		}
	}
	return true
}
func (r *siteRuntime) readers(ids any) (any, error) {
	if text(ids) == "public" {
		return "public", nil
	}
	list, err := stringsList(ids, 64, true)
	if err != nil {
		return nil, err
	}
	list = append([]string{r.owner.Public.ID}, list...)
	cards := []core.PublicIdentity{}
	seen := map[string]bool{}
	for _, id := range list {
		if seen[id] {
			continue
		}
		seen[id] = true
		if contains(r.node.config.Blocked, id) {
			return nil, errors.New("contacto bloqueado")
		}
		var card *core.PublicIdentity
		if id == r.owner.Public.ID {
			copy := r.owner.Public
			card = &copy
		} else {
			for _, c := range r.node.config.Contacts {
				if c.ID == id {
					copy := c
					card = &copy
					break
				}
			}
		}
		if card == nil {
			return nil, errors.New("adicione primeiro o cartão do destinatário")
		}
		cards = append(cards, *card)
	}
	return cards, nil
}
func (r *siteRuntime) command(body map[string]any) (any, error) {
	action := text(body["action"])
	if action == "operation" || action == "cancel" {
		if !commandShape(body, []string{"action", "name", "sequence", "operationId"}, "") {
			return nil, errors.New("comando de site inválido")
		}
		seq, err := number(body["sequence"])
		if err != nil {
			return nil, err
		}
		op, err := r.catalog.Operation(text(body["name"]), seq, text(body["operationId"]))
		if err != nil {
			return nil, err
		}
		if action == "operation" || op == nil {
			return map[string]any{"operation": op}, nil
		}
		cancelled, err := r.catalog.Cancel(text(body["name"]), op.OperationHandle)
		if err != nil {
			return nil, err
		}
		if cancelled.Phase == "cancelled" {
			r.remove(text(body["name"]), seq)
		}
		return map[string]any{"operation": cancelled}, nil
	}
	if err := r.initialize(); err != nil {
		return nil, err
	}
	if action == "state" || action == "history" {
		if !commandShape(body, []string{"action", "address"}, "") {
			return nil, errors.New("comando de site inválido")
		}
		owner, name, err := r.address(body["address"])
		if err != nil {
			return nil, err
		}
		if action == "state" {
			return r.catalog.State(owner, name)
		}
		history, err := r.catalog.History(owner, name)
		return map[string]any{"address": body["address"], "revisions": history}, err
	}
	if action == "resolve" {
		if !commandShape(body, []string{"action", "address"}, "revisionId") {
			return nil, errors.New("comando de site inválido")
		}
		_, has := body["revisionId"]
		return r.resolve(text(body["address"]), text(body["revisionId"]), has)
	}
	if action == "resume" {
		if !commandShape(body, []string{"action", "name", "sequence", "operationId"}, "") {
			return nil, errors.New("comando de site inválido")
		}
		seq, err := number(body["sequence"])
		if err != nil {
			return nil, err
		}
		op, err := r.catalog.Operation(text(body["name"]), seq, text(body["operationId"]))
		if err != nil {
			return nil, err
		}
		if op == nil {
			return nil, errors.New("operação desconhecida")
		}
		return r.finish(text(body["name"]), op.OperationHandle)
	}
	if action != "publish" || !commandShape(body, []string{"action", "name", "sequence", "operationId", "expectedBase", "payload", "recipients", "ttlMs"}, "confirmedHeads") {
		return nil, errors.New("comando de site inválido")
	}
	readers, err := r.readers(body["recipients"])
	if err != nil {
		return nil, err
	}
	request := map[string]any{"sequence": body["sequence"], "operationId": body["operationId"], "expectedBase": body["expectedBase"], "payload": body["payload"], "readers": readers, "ttlMs": body["ttlMs"]}
	if heads, exists := body["confirmedHeads"]; exists {
		request["confirmedHeads"] = heads
	}
	name := text(body["name"])
	op, err := r.catalog.CreatePublication(name, request, r.validateResources)
	if err != nil {
		return nil, err
	}
	if op.Phase == "prepared" || op.Phase == "committed" {
		r.track(name, op)
	}
	address, _ := sites.Address(r.owner.Public.ID, name)
	result, err := r.finish(name, op.OperationHandle)
	if err != nil {
		r.lastError = err.Error()
		current, e := r.current(name, op.OperationHandle)
		if e != nil {
			return nil, err
		}
		return map[string]any{"address": address, "operation": current, "error": r.lastError}, nil
	}
	result["address"] = address
	return result, nil
}
func (n *Node) cancelSitePacketsLocked() {
	if n.identity == nil {
		return
	}
	owner := n.identity.Public.ID
	n.Router.CancelLocal(func(value any) bool {
		m, ok := value.(map[string]any)
		if !ok || text(m["type"]) != "bundle" {
			return false
		}
		b, e := decodeBundle(m["bundle"])
		if e != nil || b.Manifest.Kind != "site" || b.Manifest.Author.ID != owner {
			return false
		}
		for _, k := range b.Manifest.Keys {
			if contains(n.config.Blocked, k.Reader) {
				return true
			}
		}
		return false
	})
}
